File: solutions/elixir/luhn/2/lib/luhn.ex
defmodule Luhn do
  @doc """
  Checks if the given number is valid via the luhn formula
  """
  @spec valid?(String.t()) :: boolean
  def valid?(number) do
    case valid_digits(number) do
      :invalid -> false
      digits -> sum_div_by_10?(double_every_other(digits))
    end
  end

  defp valid_digits(number, acc \\ [])
  defp valid_digits("", acc) when length(acc) < 2, do: :invalid
  defp valid_digits("", acc), do: acc
  defp valid_digits(" " <> rest, acc), do: valid_digits(rest, acc)
  defp valid_digits(<<n, rest::binary>>, acc) when n in ?0..?9 do
    valid_digits(rest, [n - ?0 | acc])
  end
  defp valid_digits(_, _), do: :invalid

  defp double_every_other(numbers, acc \\ [])
  defp double_every_other([], acc), do: acc
  defp double_every_other([n], acc), do: [n | acc]
  defp double_every_other([n, m | rest], acc) when m > 4 do
    double_every_other(rest, [m * 2 - 9, n | acc])
  end
  defp double_every_other([n, m | rest], acc) do
    double_every_other(rest, [m * 2, n | acc])
  end

  defp sum_div_by_10?(numbers) do
    rem(Enum.sum(numbers), 10) == 0
  end
end

File: solutions/erlang/luhn/2/luhn.erl
-module(luhn).
-export([valid/1, create/1]).

valid(Digits) ->
  Number = to_int_list(Digits),
  Adjusted = adjust(Number, []),
  0 =:= sum(Adjusted) rem 10.

create(Digits) ->
  Number = [0 | to_int_list(Digits)],
  Adjusted = adjust(Number, []),
  CheckDigit = 10 - (sum(Adjusted) rem 10),
  Digits ++ integer_to_list(CheckDigit).

%% Strips non-digits and reverses the value to feed into adjust/2.
to_int_list(Digits) ->
  [D - $0 || D <- lists:reverse(Digits), D >= $0 andalso D =< $9].

%% Applies double_ish/1 to every second digit.
adjust([], Acc) ->
  Acc;
adjust([X, Y | Rest], Acc) ->
  adjust(Rest, [double_ish(Y), X | Acc]);
adjust([X], Acc) ->
  [X | Acc].

%% Double the digit, but subtract 9 if 10 or greater.
double_ish(N) when N =< 4 ->
  N * 2;
double_ish(N) ->
  N * 2 - 9.

sum(IntList) ->
  lists:foldl(fun(E, A) -> A + E end, 0, IntList).